Emit app port and added get*Version() functions.

ddd-mtl opened the pull request:
Reads the `APP_WS_PORT:` output from holochain runner and emit value back to client.
The code makes sure IS_READY and app port has been emitted before resolving promise.

ddd-mtl committed:
Handle uid and bootstrap parameter.
Emit app port
Added get*Version functions.

ddd-mtl committed:
Merge branch 'main' of https://github.com/Sprillow/electron-holochain

# Resolved Conflicts:
#	src/options.ts

ddd-mtl committed:
bootstrap typo fix

Connoropolous commented:
nice to have your contributions here @ddd-mtl , these are things I've been meaning to tighten up!

wouldn't it have to have waited for app ws to be ready anyways? it wasn't working like that?

Connoropolous commented:
I will note that I'm aware that handling when holochain crashes for various reasons is hard and hasn't been yet handled nicely in this library

ddd-mtl commented:
Once an app is installed, the runner doesn't care about the appWs option you give it. So I need to retrieve whatever appWs its currently using and tell my web ui to use that.
Similar thing with the appId. The runner just gives you the first app installed it finds, whatever option you feed it.

Connoropolous commented:
"Once an app is installed, the runner doesn't care about the appWs option you give it. So I need to retrieve whatever appWs its currently using and tell my web ui to use that." this is true, though if you told it to use that port, isn't it safe to assume that you know what the port is?

Connoropolous commented on src/options.ts:
```diff
@@ -27,8 +27,8 @@ export function constructOptions(options: HolochainRunnerOptions): string[] {
   }
-  if (options.boostrapUrl) {
-    optionsArr = optionsArr.concat(['--bootstrap-url', options.boostrapUrl])
+  if (options.bootstrapUrl) {
```
oh, shoot. my bad!

ddd-mtl commented:
> "Once an app is installed, the runner doesn't care about the appWs option you give it. So I need to retrieve whatever appWs its currently using and tell my web ui to use that." this is true, though if you told it to use that port, isn't it safe to assume that you know what the port is?

Probably but you need to remember what port you told it to use the first time.
Also the wanted port could already be taken. Haven't tried it but at best it gives back a different port, worst case it crashs.

Connoropolous commented:
Yah... I can attest that in the case where the port is already taken holochain crashes

Connoropolous commented:
I dont think there can be any harm in listening for the port it is using, but I know that its a guarantee that it'll be the port you asked for. 

I suppose the exception is passing in a 0. I do not know if this is possible. In that case it could (but does it?) give you back an open port that it attached to.

ddd-mtl commented:
Don't know with the runner, but thats what I do with Snapmail. I give holochain a 0 for app port and write in the conductor-config the port actually used (for future use). 
Useful for running multiple instance of your happ.

Connoropolous commented:
well, we do just call `add_app_interface` for the conductor, so it's just a passthrough, however, I am noting about that code that it would not display the value picked by the conductor as the APP_WS_PORT, but the value passed in, such as `0`
https://github.com/Sprillow/holochain-runner/blob/eb9631e5568c3668d7681fe2b10f570464db3a31/src/embedded_runner.rs#L193-L204
we could easily adjust that. I am assuming that `add_app_interface` does return the selected port

Connoropolous commented:
Actually... I am wondering about this: 
"I give holochain a 0 for app port and write in the conductor-config the port actually used (for future use)."
I just looked in holochain source code and can't see any evidence of app port accepting a `0` value. 

admin port does. Is that what you were referring to in your above quote?

ddd-mtl committed:
Fixed bug with initAgent() that would not register 'will-quit' callback if holochain output an error message.
Changed kill routine to also kill grand-children.
Safety checks on get*Version()

ddd-mtl commented:
I just tried it out:

If I give `appWsPort: 0` it will connect to `8888`
If I give `appWsPort: 1` it will connect to `1` even though it doesn't work.

Connoropolous commented:
Can you tell me a bit about what you did with error handling in the latest commit? I would like to understand it quite well, and document it

ddd-mtl committed:
Bumped holochain_runner version used.
Added app port event to readme

ddd-mtl committed:
comments for killHolochain

Connoropolous commented:
published to npm as v0.0.25
